packages/internal/src/kv-store.js

```diff
@@ -0,0 +1,263 @@
+   * @returns {boolean} true if a value is stored for the key, false if not.
+   * @throws if key is not a string.
+   */
+  function has(key) {
+    typeof key === 'string' || Fail`key must be a string`;
+    return get(key) !== undefined;
+  }
+
+  const sqlKVSet = db.prepare(`
+    INSERT INTO kvStore (key, value)
+    VALUES (?, ?)
+    ON CONFLICT DO UPDATE SET value = excluded.value
+  `);
+
+  /**
+   * Store a value for a given key. The value will replace any prior value if
+   * there was one.
+   *
+   * @param {string} key The key whose value is being set.
+   * @param {string} value The value to set the key to.
+   * @throws if either parameter is not a string.
+   */
+  function set(key, value) {
+    typeof key === 'string' || Fail`key must be a string`;
+    typeof value === 'string' || Fail`value must be a string`;
+    // synchronous read after write within a transaction is safe
+    // The transaction's overall success will be awaited during commit
+    beforeMutation();
+    sqlKVSet.run(key, value);
+    trace('set', key, value);
+  }
+
+  const sqlKVDel = db.prepare(`
+    DELETE FROM kvStore
+    WHERE key = ?
+  `);
+
+  /**
+   * Remove any stored value for a given key. It is permissible for there to be
+   * no existing stored value for the key.
+   *
+   * @param {string} key The key whose value is to be deleted
+   * @throws if key is not a string.
+   */
+  function del(key) {
+    typeof key === 'string' || Fail`key must be a string`;
+    beforeMutation();
+    sqlKVDel.run(key);
+    trace('del', key);
+  }
+
+  const kvStore = {
+    has,
+    get,
+    getNextKey,
+    set,
+    delete: del,
+  };
+
+  return kvStore;
+}
+
+// TODO: Replace compareByCodePoints and makeKVStoreFromMap and
+// provideEnhancedKVStore with imports when
+// available.
+// https://github.com/Agoric/agoric-sdk/pull/10299
+
+export const compareByCodePoints = (left, right) => {
+  const leftIter = left[Symbol.iterator]();
+  const rightIter = right[Symbol.iterator]();
+  for (;;) {
+    const { value: leftChar } = leftIter.next();
+    const { value: rightChar } = rightIter.next();
+    if (leftChar === undefined && rightChar === undefined) {
+      return 0;
+    } else if (leftChar === undefined) {
+      // left is a prefix of right.
+      return -1;
+    } else if (rightChar === undefined) {
+      // right is a prefix of left.
+      return 1;
+    }
+    const leftCodepoint = /** @type {number} */ (leftChar.codePointAt(0));
+    const rightCodepoint = /** @type {number} */ (rightChar.codePointAt(0));
+    if (leftCodepoint < rightCodepoint) return -1;
+    if (leftCodepoint > rightCodepoint) return 1;
+  }
+};
+
+/**
+ * @template [T=unknown]
+ * @typedef {{
+ *   has: (key: string) => boolean;
+ *   get: (key: string) => T | undefined;
+ *   getNextKey: (previousKey: string) => string | undefined;
+ *   set: (key: string, value: T) => void;
+ *   delete: (key: string) => void;
+ * }} KVStoreFromMap
+ */
```

Comment:
I don't see any benefit to a KVStoreFromMap type; let's just use a single type defaulting to string values.
```suggestion
/**
 * @template [T=string]
 * @typedef {{
 *   has: (key: string) => boolean;
 *   get: (key: string) => T | undefined;
 *   getNextKey: (previousKey: string) => string | undefined;
 *   set: (key: string, value: T) => void;
 *   delete: (key: string) => void;
 * }} KVStore
 */
```